### PR TITLE
Extra attributes for patching convenience

### PR DESCRIPTION
**Pull request by SokyranTheDragon**
I've thought something like this could potentially be useful so I decided to try and quickly make something small... Aaaand, here we are a day later with Over 400 lines of extra code. Oh well.

This is not something that's fully required for MP Compat, nor something I really need myself. Just small stuff convenience. So I'd say @notfood it's up to you if it's something you'd want to include in, or if you'd rather leave it out.

Also, let me know if there's any changes you'd want me to do here as well.

Changes/additions:

- Created `MpCompatAttributes` file
- Moved `MpCompatFor` and `MpCompatRequireMod` attributes to the new file to keep all attributes together
- Added abstract `MpCompatPatch` attribute with subclasses `MpCompat(Prefix/Postfix/Transpiler/Finalizer)` for convenience of patching (they are based on equivalent attributes from Multiplayer, with some additions and inclusion of the finalizer)
- Added `MpCompatSync(Method/Field/Worker)` attributes for convenience of registering them
- Added static class `MpCompatPatchLoader` which `LoadPatch` method that will go through all methods/fields in a single type and apply Harmony patches and register Multiplayer sync methods/fields/workers marked by the earlier mentioned attributes
- Modified Simple Sidearms and Vanilla Factions Expanded - Ancients patches to use those new features, presenting an example usage

Currently, the patches are applied manually by the compat itself. I've considered an alternative approach by having `MpCompatLoader` try and apply those patches for all active compats, however - this ended up being a very messy solution and was adding extra complexity to loading, so I decided to drop this approach.

### CODE REVIEW

File: Source/Mods/SimpleSidearms.cs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ﻿using System;
 
 using HarmonyLib;
+using JetBrains.Annotations;
```
Comment (notfood) [inline Source/Mods/SimpleSidearms.cs:4]: How's JetBrains.Annotations sneaking in? Don't you need it in the csproj?

File: Source/MpCompatAttributes.cs [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+                        }
+                    }
+                    catch (Exception e)
+                    {
+                        Log.Error($"MpCompatPatch {method.DeclaringType}.{method.Name} failed with exception: {e}");
+                    }
+                }
+            }
+
+            foreach (var field in AccessTools.GetDeclaredFields(type).Where(f => f.IsStatic && typeof(ISyncField).IsAssignableFrom(f.FieldType)))
```
Comment (notfood) [inline Source/MpCompatAttributes.cs:139]: Can't this be merged above? It's calling the same `AccessTools.GetDeclaredFields(type).Where(m => m.IsStatic)`

**Review by notfood — COMMENTED** [review]
Overall I like it, simplifies things that are repeated often in MPCompat. The LoadPatch method is kind of hard to read but I get the intention.

File: Source/Mods/SimpleSidearms.cs (hunk 1 above)
Comment (SokyranTheDragon) [reply Source/Mods/SimpleSidearms.cs:4]: `UnityEngine.CoreModule` includes them, which we've got through the `Krafs.Rimworld.Ref` NuGet package.

File: Source/MpCompatAttributes.cs (hunk 2 above)
Comment (SokyranTheDragon) [reply Source/MpCompatAttributes.cs:139]: It's not `GetDeclaredFields` above, but `GetDeclaredMethods`.

**Comment by SokyranTheDragon** [comment]
I could try to simplify the `LoadPatch` method, but I'm not really sure how besides spiting each case into a separate method (so the `LoadPatch` itself should be easier to read, but the other methods may not be as clear).

**Comment by SokyranTheDragon** [comment]
As a side not, the `LoadPatch` method was somewhat based on those 2 methods from MP itself:
https://github.com/rwmt/Multiplayer/blob/master/Source/Client/Util/MpPatch.cs#L105-L139
https://github.com/rwmt/Multiplayer/blob/d03a1b8f50c66b12f3b6182da359c5d4a80355cf/Source/Client/Syncing/Sync.cs#L128-L169

They obviously could not just be copy-pasted, but served as an inspiration.

**Comment by notfood** [comment]
It's fine

**Pull request merged**